master/pkg/model/experiment.go

```diff
@@ -403,6 +408,20 @@ type TrialLog struct {
 	StdType     *string    `db:"stdtype" json:"stdtype"`
 }
 
+// Proto converts a trial log to its protobuf representation.
+func (t TrialLog) Proto() *apiv1.TrialLogsResponse {
+	resp := &apiv1.TrialLogsResponse{Message: t.Message}
+	switch {
+	case t.ID != nil:
+		resp.IdOneof = &apiv1.TrialLogsResponse_Id{Id: int64(*t.ID)}
+	case t.StringID != nil:
+		resp.IdOneof = &apiv1.TrialLogsResponse_StringId{StringId: *t.StringID}
+	default:
+		panic("log had no valid ID")
```

Comment:
*sigh* I've been picking up Rust lately; I think I've realized that all I ever wanted was a systems language with real union/enum types and pattern matching.

Reply:
Lol, discriminated unions are the best, I agree. With generics, you can write a package for them at least.